mypy/checker.py

```diff
@@ -1832,10 +1846,18 @@ def check_multi_assignment_from_union(self, lvalues: List[Expression], rvalue: E
             # Bind a union of types collected in 'assignments' to every expression.
             if isinstance(expr, StarExpr):
                 expr = expr.expr
-            types, declared_types = zip(*items)
+
+            # TODO: See todo in binder.py, ConditionalTypeBinder.assign_type
+            # It's unclear why the 'declared_type' param is sometimes 'None'
+            clean_items = []  # type: List[Tuple[Type, Type]]
+            for type, declared_type in items:
+                assert declared_type is not None
```

Comment:
I think what was happening here was that previously, the call to `zip` directly below was incorrectly inferred to have a type of 'Any', which made the call to `self.binder.assign_type` always succeed.

My overload changes made zip start inferring the correct types here, which led to a type error when calling `binder.assign_type`. I thought about adding casts, but was wanted to avoid silently propagating 'None' values into unexpected places as much as possible: I figured a noisy assert would be easier for others to debug if this ever became an issue.

That said, mypy isn't running into any issues here so maybe a cast is safe: let me know and I can make that change.

Reply:
OK, let's keep assert for a moment, I will run `testpr` on our internal codebases.